File: src/nostr/nip_01.rs
use crate::nostr::{Event, Filter};
use serde_json::value::RawValue;
use serde::Deserialize;

#[derive(Debug, Clone)]
pub enum ClientMessage {
    Event(Event),
    Req(String, Vec<Filter>),
    Close(String),
}

#[derive(Deserialize)]
struct PartialEvent {
    id: String,
}

pub enum PartialClientMessage {
    Event(String), // Returns the ID if it looks like an event
}

impl PartialClientMessage {
    pub fn from_json(text: &str) -> Option<Self> {
        let arr: Vec<&RawValue> = serde_json::from_str(text).ok()?;
        if arr.len() < 2 { return None; }

        let msg_type: String = serde_json::from_str(arr[0].get()).ok()?;
        if msg_type == "EVENT" {
            let event: PartialEvent = serde_json::from_str(arr[1].get()).ok()?;
            return Some(Self::Event(event.id));
        }
        None
    }
}

impl ClientMessage {
    pub fn from_json(text: &str) -> Result<Self, String> {
        let arr: Vec<&RawValue> = serde_json::from_str(text).map_err(|e| e.to_string())?;

        if arr.is_empty() {
            return Err("empty message".to_string());
        }

        let msg_type: String = serde_json::from_str(arr[0].get()).map_err(|e| e.to_string())?;

        match msg_type.as_str() {
            "EVENT" if arr.len() >= 2 => {
                let event: Event = serde_json::from_str(arr[1].get()).map_err(|e| e.to_string())?;
                Ok(Self::Event(event))
            }
            "REQ" if arr.len() >= 3 => {
                let sub_id: String = serde_json::from_str(arr[1].get()).map_err(|e| e.to_string())?;
                let mut filters = Vec::new();
                for value in &arr[2..] {
                    let filter: Filter = serde_json::from_str(value.get()).map_err(|e| e.to_string())?;
                    filters.push(filter);
                }
                Ok(Self::Req(sub_id, filters))
            }
            "CLOSE" if arr.len() >= 2 => {
                let sub_id: String = serde_json::from_str(arr[1].get()).map_err(|e| e.to_string())?;
                Ok(Self::Close(sub_id))
            }
            v => Err(format!("unknown message type: {}", v)),
        }
    }
}

#[derive(Debug, Clone, PartialEq, Eq)]
pub enum RelayMessage {
    Ok(String, bool, String),
    Event(String, crate::nostr::Event),
    Eose(String),
    Notice(String),
    Closed(String, String),
}

impl RelayMessage {
    pub fn from_json(text: &str) -> Result<Self, String> {
        let arr: Vec<&RawValue> = serde_json::from_str(text).map_err(|e| e.to_string())?;

        if arr.is_empty() {
            return Err("empty message".to_string());
        }

        let msg_type: String = serde_json::from_str(arr[0].get()).map_err(|e| e.to_string())?;

        match msg_type.as_str() {
            "OK" if arr.len() >= 4 => {
                let id: String = serde_json::from_str(arr[1].get()).map_err(|e| e.to_string())?;
                let ok: bool = serde_json::from_str(arr[2].get()).map_err(|e| e.to_string())?;
                let msg: String = serde_json::from_str(arr[3].get()).map_err(|e| e.to_string())?;
                Ok(Self::Ok(id, ok, msg))
            }
            "EVENT" if arr.len() >= 3 => {
                let sub_id: String = serde_json::from_str(arr[1].get()).map_err(|e| e.to_string())?;
                let event: crate::nostr::Event = serde_json::from_str(arr[2].get()).map_err(|e| e.to_string())?;
                Ok(Self::Event(sub_id, event))
            }
            "EOSE" if arr.len() >= 2 => {
                let sub_id: String = serde_json::from_str(arr[1].get()).map_err(|e| e.to_string())?;
                Ok(Self::Eose(sub_id))
            }
            "NOTICE" if arr.len() >= 2 => {
                let msg: String = serde_json::from_str(arr[1].get()).map_err(|e| e.to_string())?;
                Ok(Self::Notice(msg))
            }
            "CLOSED" if arr.len() >= 3 => {
                let sub_id: String = serde_json::from_str(arr[1].get()).map_err(|e| e.to_string())?;
                let msg: String = serde_json::from_str(arr[2].get()).map_err(|e| e.to_string())?;
                Ok(Self::Closed(sub_id, msg))
            }
            v => Err(format!("unknown message type: {}", v)),
        }
    }

    pub fn to_json(&self) -> String {
        match self {
            Self::Ok(id, ok, msg) => serde_json::json!(["OK", id, ok, msg]).to_string(),
            Self::Event(sub_id, event) => serde_json::json!(["EVENT", sub_id, event]).to_string(),
            Self::Eose(sub_id) => serde_json::json!(["EOSE", sub_id]).to_string(),
            Self::Notice(msg) => serde_json::json!(["NOTICE", msg]).to_string(),
            Self::Closed(sub_id, msg) => serde_json::json!(["CLOSED", sub_id, msg]).to_string(),
        }
    }
}

#[cfg(test)]
#[path = "nip_01_test.rs"]
mod nip_01_test;

File: src/nostr/nip01.ts
import { Tag } from './tag.ts';
import { Event } from './event.ts';
import { Filter, filterFromJSON } from './filter.ts';

export type ClientMessage =
  | { type: 'EVENT'; event: Event }
  | { type: 'REQ'; subscriptionId: string; filters: Filter[] }
  | { type: 'CLOSE'; subscriptionId: string };

export type PartialClientMessage =
  | { type: 'EVENT'; id: string };

export function parsePartialClientMessage(text: string): PartialClientMessage | null {
  try {
    const arr = JSON.parse(text) as unknown[];
    if (arr.length < 2) return null;
    if (arr[0] === 'EVENT') {
      const event = arr[1] as Record<string, unknown>;
      if (typeof event?.id === 'string') {
        return { type: 'EVENT', id: event.id };
      }
    }
    return null;
  } catch {
    return null;
  }
}

export function parseClientMessage(text: string): ClientMessage {
  const arr = JSON.parse(text) as unknown[];
  if (arr.length === 0) throw new Error('empty message');

  const msgType = arr[0];
  if (typeof msgType !== 'string') throw new Error(`unknown message type: ${msgType}`);

  switch (msgType) {
    case 'EVENT': {
      if (arr.length < 2) throw new Error(`unknown message type: ${msgType}`);
      const event = parseEventJSON(arr[1] as Record<string, unknown>);
      return { type: 'EVENT', event };
    }
    case 'REQ': {
      if (arr.length < 3) throw new Error(`unknown message type: ${msgType}`);
      const subId = arr[1];
      if (typeof subId !== 'string') throw new Error(`unknown message type: ${msgType}`);
      const filters = (arr.slice(2) as Record<string, unknown>[]).map(f => filterFromJSON(f));
      return { type: 'REQ', subscriptionId: subId, filters };
    }
    case 'CLOSE': {
      if (arr.length < 2) throw new Error(`unknown message type: ${msgType}`);
      const subId = arr[1];
      if (typeof subId !== 'string') throw new Error(`unknown message type: ${msgType}`);
      return { type: 'CLOSE', subscriptionId: subId };
    }
    default:
      throw new Error(`unknown message type: ${msgType}`);
  }
}

export type RelayMessage =
  | { type: 'OK'; id: string; ok: boolean; message: string }
  | { type: 'EVENT'; subscriptionId: string; event: Event }
  | { type: 'EOSE'; subscriptionId: string }
  | { type: 'NOTICE'; message: string }
  | { type: 'CLOSED'; subscriptionId: string; message: string };

export function parseRelayMessage(text: string): RelayMessage {
  const arr = JSON.parse(text) as unknown[];
  if (arr.length === 0) throw new Error('empty message');

  const msgType = arr[0];
  if (typeof msgType !== 'string') throw new Error(`unknown message type: ${msgType}`);

  switch (msgType) {
    case 'OK': {
      if (arr.length < 4) throw new Error('OK requires id, bool, message');
      return {
        type: 'OK',
        id: arr[1] as string,
        ok: arr[2] as boolean,
        message: arr[3] as string,
      };
    }
    case 'EVENT': {
      if (arr.length < 3) throw new Error('EVENT requires sub_id and event');
      return {
        type: 'EVENT',
        subscriptionId: arr[1] as string,
        event: parseEventJSON(arr[2] as Record<string, unknown>),
      };
    }
    case 'EOSE': {
      if (arr.length < 2) throw new Error('EOSE requires sub_id');
      return { type: 'EOSE', subscriptionId: arr[1] as string };
    }
    case 'NOTICE': {
      if (arr.length < 2) throw new Error('NOTICE requires message');
      return { type: 'NOTICE', message: arr[1] as string };
    }
    case 'CLOSED': {
      if (arr.length < 3) throw new Error('CLOSED requires sub_id and message');
      return {
        type: 'CLOSED',
        subscriptionId: arr[1] as string,
        message: arr[2] as string,
      };
    }
    default:
      throw new Error(`unknown message type: ${msgType}`);
  }
}

export function relayMessageToJSON(msg: RelayMessage): string {
  switch (msg.type) {
    case 'OK':
      return JSON.stringify(['OK', msg.id, msg.ok, msg.message]);
    case 'EVENT':
      return JSON.stringify(['EVENT', msg.subscriptionId, msg.event]);
    case 'EOSE':
      return JSON.stringify(['EOSE', msg.subscriptionId]);
    case 'NOTICE':
      return JSON.stringify(['NOTICE', msg.message]);
    case 'CLOSED':
      return JSON.stringify(['CLOSED', msg.subscriptionId, msg.message]);
  }
}

function parseEventJSON(json: Record<string, unknown>): Event {
  return {
    id: json.id as string,
    pubkey: json.pubkey as string,
    createdAt: json.created_at as number,
    kind: json.kind as number,
    tags: ((json.tags as Array<Array<unknown>>) ?? []).map(t => Tag.fromJSON(t as never)),
    content: json.content as string,
    sig: json.sig as string,
  };
}
